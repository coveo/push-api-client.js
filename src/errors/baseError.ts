export abstract class PushApiClientBaseError extends Error {
  public name = 'Push API Client Error';
  public constructor(message?: string) {
    super(message);
  }
}
