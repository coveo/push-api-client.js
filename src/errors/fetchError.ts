import {PushApiClientBaseError} from './baseError';
import dedent from 'ts-dedent';

export class FetchError extends PushApiClientBaseError {
  public name = 'Fetch Error';
  public static async build(response: Response) {
    let body: string;
    try {
      body = JSON.stringify(await response.json());
    } catch (error) {
      body = await response.text();
    }
    return new FetchError(body);
  }
  private constructor(public body: string) {
    super();
    this.message = dedent`
      Response body:
      ${body}
    `;
  }
}
