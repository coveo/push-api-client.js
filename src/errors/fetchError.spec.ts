import type {Response} from 'undici';
import {PushApiClientBaseError} from './baseError';
import {FetchError} from './fetchError';

describe('FetchError', () => {
  const fakeResponseJson = jest.fn();
  const fakeResponseText = jest.fn();
  const fakeResponse: Response = {
    json: fakeResponseJson,
    text: fakeResponseText,
  } as unknown as Response;

  beforeEach(() => {
    jest.resetAllMocks();
    fakeResponseJson.mockResolvedValue(null);
    fakeResponseText.mockResolvedValue('hello there');
  });

  it('has a proper name', async () => {
    expect((await FetchError.build(fakeResponse)).name).toBe('Fetch Error');
  });

  it('inherit the PushApiClientBaseError', async () => {
    expect(await FetchError.build(fakeResponse)).toBeInstanceOf(
      PushApiClientBaseError
    );
  });

  describe('when json returns a stringifiable object', () => {
    beforeEach(() => {
      fakeResponseJson.mockResolvedValueOnce({hello: 'world'});
    });

    it('uses its stringify value in the message', async () => {
      expect((await FetchError.build(fakeResponse)).message).toMatchSnapshot();
    });
  });

  describe('when json returns a unstringifiable object', () => {
    beforeEach(() => {
      const cyclicObject: {self?: object} = {};
      cyclicObject.self = cyclicObject;
      fakeResponseJson.mockResolvedValueOnce(cyclicObject);
    });

    it('uses the text value in the message', async () => {
      expect((await FetchError.build(fakeResponse)).message).toMatchSnapshot();
    });
  });

  describe('when json and text fails', () => {
    beforeEach(() => {
      fakeResponseJson.mockRejectedValueOnce('nope');
      fakeResponseText.mockRejectedValueOnce('nope');
    });

    it('does not returns a FetchError', async () => {
      await expect(FetchError.build(fakeResponse)).rejects.not.toBeInstanceOf(
        FetchError
      );
    });
  });
});
