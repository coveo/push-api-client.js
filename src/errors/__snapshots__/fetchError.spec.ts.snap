// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`FetchError when json returns a stringifiable object uses its stringify value in the message 1`] = `
"Response body:
{"hello":"world"}"
`;

exports[`FetchError when json returns a unstringifiable object uses the text value in the message 1`] = `
"Response body:
hello there"
`;
